3단계 - 경로 조회 기능

hbkuk opened the pull request:
안녕하세요. 승철 리뷰어님. 😄 

2주차 `3단계 - 경로 조회 기능` 미션에 대한 코드 리뷰 요청드립니다.

이번 미션은 `Outside In TDD` 방식으로 구현해 보았습니다.

컨트롤러를 만들고 난 후, 경로 조회를 도대체 어떻게 구현해야 할지 막막했었는데요.
비즈니스 플로우를 만들겸 `Path Finder` 객체를 Stubbing 해서 , 서비스 레이어의 단위 테스트를 작성하다보니 ... 빠르게 기능을 구현할 수 있었습니다.

--- 

지난 번에 `sortByConnectedSections` 리팩토링을 어떻게 해야할지 조언을 부탁드렸었는데요.
최대한으로, 리팩토링을 진행해 보았습니다. 괜찮은 방향으로 진행했는지.. 확인부탁드립니다. 😄 

또, 코멘트 달아주신 내용에 대해서 고민이 되는 부분이 있습니다. 확인 부탁드리겠습니다.(https://github.com/next-step/atdd-subway-path/pull/770) 😢

hbkuk committed:
refactor: 복잡한 sortByConnectSections 내부 로직을 작은 단위로 분리

- fix: 사용하지 않은 매개변수 삭제

- fix: 잘못 변경된 접근 제어자 수정

hbkuk committed:
refactor: 인수 테스트에서 상태 코드 검증은 구간 추가 혹은 삭제 요청 메서드에서 진행되도록 수정

hbkuk committed:
refactor: Section Steps 리팩토링

hbkuk committed:
test: 지하철 경로 조회 요구사항에 대한 인수 테스트 추가

hbkuk committed:
test: 지하철 경로 조회 요구사항에 대한 컨트롤러 구현

hbkuk committed:
test: 지하철 경로 조회 요구사항에 대한 서비스 레이어의 단위 테스트 추가

hbkuk committed:
feat: 지하철 경로 조회 관련 서비스 레이어의 단위 테스트에 만족하는 기능 구현

hbkuk committed:
test: 경로 조회 기능 구현을 위한 경로 조회 컴포넌트에 대한 단위 테스트 추가

hbkuk committed:
feat: 경로 조회 기능 구현을 위한 경로 조회 컴포넌트에 만족하는 기능 구현

hbkuk committed:
test: 지하철 경로 조회 요구사항 관련 인수 테스트에 만족하는 기능 수정

hbkuk committed:
test: 지하철 경로 조회 요구사항에 대한 서비스 레이어의 단위 테스트 수정

hbkuk committed:
refactor: 경로 조회 컴포넌트(Path Finder) 메서드 리팩토링

hbkuk committed:
refactor: 경로 조회 컴포넌트에 대한 단위 테스트 리팩토링

hbkuk committed:
refactor: 지하철 경로 조회 요구사항에 대한 서비스 레이어의 단위 테스트 리팩토링

hbkuk committed:
refactor: 지하철 경로 조회 요구사항에 대한 인수 테스트 리팩토링

- feat: 출발역과 도착역 정보를 담는 DTO 객체 추가

hbkuk committed:
test: 지하철 경로 조회 요구사항 관련 Controller, Service 레이어 리팩토링

hbkuk committed:
fix: 기존 Station Fixture 수정 및 관련 로직 수정

hbkuk committed:
refactor: 지하철 경로 조회 요구사항에 대한 인수 테스트에서 사용된 Fixture 리팩토링

hbkuk committed:
refactor: 지하철 경로 조회 요구사항에 대한 서비스 레이어의 단위 테스트 리팩토링

hbkuk committed:
refactor: 경로 조회 컴포넌트에 대한 단위 테스트 리팩토링

hbkuk committed:
test: 경로 조회 컴포넌트 관련 Side 케이스 추가

hbkuk committed:
feat: 경로 조회 컴포넌트 관련 Side 케이스에 대한 기능 구현

hbkuk committed:
docs: 불필요한 <p> 태그 삭제

hbkuk committed:
fix: 경로 조회 컴포넌트 관련 테스트 클래스와 서비스 레이어 관련 테스트 클래스에 설정된 @BeforeEach 메서드 리팩토링

pch8388 commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -108,30 +111,39 @@ private boolean isLastStation(Station stationToDelete) {
 
+    private void connectBasedOnSection(Section baseSection) {
+        for (int indexToConnect = sections.indexOf(baseSection); indexToConnect < sections.size(); indexToConnect++) {
+            if (canConnect(baseSection, sections.get(indexToConnect))) {
+                break;
+            }
```
[객체지향 생활체조](https://edu.nextstep.camp/s/vLaUvWbK/ls/U4Y1nOBw)
 > 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.

hbkuk replied:
🙄 
(호다닥 수정해보겠습니다.)

pch8388 commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -108,30 +111,39 @@ private boolean isLastStation(Station stationToDelete) {
-                    sections.add(baseIndex + 1, sectionToConnect);
-                    break;
-                }
-            }
+            connectBasedOnSection(sections.get(baseIndex));
```
수정해주신 부분들이 많이 깔끔해졌다고 느끼는 데요 😄 
다만 collection 을 index 로 다루다보면 실수할 여지가 많은 것 같아요 
sections 를 정렬을 위해 수정하는 것 또한 불안 요소가 늘어난 느낌입니다 😢 
추천해드리고 싶은 방향은 sections 는 그래도 두고, 정렬 로직을 구현한 `getAllSections` 같은 것을 호출하여 사용하는 방법입니다. 😄 
사실 잘생각해보면 구간 추가나 삭제 시에 정렬되어있으면 편하지만 꼭 정렬되있어야 실행할 수 있는 로직은 아니거든요 🤔 

구간 추가 : 구간을 추가할 위치를 찾는다. 구간을 추가할 위치는 제일 처음 / 제일 끝 / 추가할 상행역 또는 하행역이 일치하는 구간 사이이다. 

와 같이 반드시 정렬되어있지 않아도 된다고 생각해요 😄

hbkuk replied:
역시.. 가려웠던 부분을 시원하게 긁어주시네요. 😆 

최대한 고민하고 반영해보겠습니다 ~! 👍

pch8388 commented on src/main/java/nextstep/subway/presentation/LineController.java:
```diff
@@ -46,4 +48,10 @@ public ResponseEntity<Void> deleteLine(@PathVariable Long lineId) {
         return ResponseEntity.noContent().build();
     }
+
+    @GetMapping("/paths")
```
어울리지 않는 느낌인데 여기에 두신 이유가 있을까요? 🤔

hbkuk replied:
흠 ... 이부분은 애초에 설계를 할때부터 꼬인 것 같아요.

경로 조회 요구사항에 대해서 `LineServiceMockTest.java`에 서비스 레이어 테스트 코드를 만들다보니, 
'`PathService` 라는 서비스 객체를 만들었으면 어땠을까?' 라는 생각이 듭니다.

아래 언급하신 내용과 이어지는 부분인 것 같습니다.
> 의존성이 늘어날수록 테스트는 힘들어질수 있습니다 😢

---
```
// LineServiceMockTest.java
    @Mock
    PathFinder pathFinder;
```

위와 같이 가짜 객체를 추가하면서 뭔가 찝찝했습니다. 

설계에 신경을 쓴 후, 테스트를 작성하는 것이 더 좋을 방향이었을 것인데 .. 
(요즘 미션을 진행하다 보니, 저만 너무 느린 것 같아서 심적 여유가 1도 없었습니다.😢)

우선, 지금이라도 저의 문제를 확인했으니깐, 다음부터는 조금 더 여유를 가지고 진행해보겠습니다.

pch8388 replied:
미션을 수행하시면서 늦었다라고 생각하면서 포기하는 분도 많은 것 같아요 😢 
조금 느리더라도 더 많은걸 얻어 갈 수 있도록 고민을 많이 해보시면 좋겠습니다 🔥

pch8388 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -72,6 +76,35 @@ public void deleteSection(Long lineId, Long stationIdToDelete) {
+        }
+    }
+
+    private boolean existsStation(Long stationId) {
+        return stationRepository.existsById(stationId);
```
`findStation`에서 결국 조회를 다시 하는 데, 이 구문이 꼭 필요할까요? 🤔

hbkuk replied:
`findShortestPath`가 호출되고 역 번호가 없을 때, 고유한 예외 메시지를 전달하고자 했었는데.. 

다시 생각해보니, 그다지 큰 이슈는 아닌 것 같아 `findStation`예외 메시지를 사용하도록 수정해보겠습니다~!

pch8388 commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -130,7 +163,7 @@ private StationResponse convertToStationResponse(Station station) {
     }
 
-    public SectionResponse convertToSectionResponse(Section section) {
+    private SectionResponse convertToSectionResponse(Section section) {
```
convert 로직이 굉장히 많이 생기는데, 클래스를 하나 생성하는 건 어떨까요? 😄

hbkuk replied:
오 좋은 것 같습니다. 👍

pch8388 commented on src/test/java/nextstep/subway/unit/PathFinderComponentTest.java:
```diff
@@ -0,0 +1,215 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+@ApplicationTest
```
이 애노테이션이 현재 테스트에 필요한가요?

hbkuk replied:
🙄 (삭제하겠습니다.)

pch8388 commented on src/test/java/nextstep/subway/unit/PathFinderComponentTest.java:
```diff
@@ -0,0 +1,215 @@
+    void 사전_노선_설정() {
+
+        이호선 = new Line("이호선", "green");
+        ReflectionTestUtils.setField(이호선, "id", 1L);
```
PathFinder 가 노선id 를 사용하나요?

hbkuk replied:
🙄 (삭제하겠습니다.)

pch8388 commented on src/test/java/nextstep/subway/unit/LineServiceMockTest.java:
```diff
@@ -40,75 +39,222 @@ public class LineServiceMockTest {
     @Mock
     LineRepository lineRepository;
 
+    @Mock
+    PathFinder pathFinder;
```
의존성이 늘어날수록 테스트는 힘들어질수 있습니다 😢

hbkuk replied:
미션을 진행하면서, 어떤 의미인지 정확히 느꼈습니다. 😢

pch8388 commented on src/test/java/nextstep/subway/unit/LineServiceMockTest.java:
```diff
@@ -40,75 +39,222 @@ public class LineServiceMockTest {
+        @Test
+        void 강남역에서_남부터미널역까지_경로_조회() {
+            // given
+            when(stationRepository.existsById(강남역_번호)).thenReturn(true);
+            when(stationRepository.existsById(남부터미널역_번호)).thenReturn(true);
+            when(stationRepository.findById(강남역_번호)).thenReturn(Optional.of(강남));
+            when(stationRepository.findById(남부터미널역_번호)).thenReturn(Optional.of(남부터미널));
+            when(lineRepository.findAll()).thenReturn(모든_노선_목록);
+            when(pathFinder.calculateShortestPath(모든_노선_목록, 강남, 남부터미널)).thenReturn(new PathResult(List.of(강남, 교대, 남부터미널), 12));
```
mockito 에 대한 의존 영역이 점점 넓어지는 것 같네요 😢 
[참고](https://kukim.tistory.com/107) 글을 읽어보고 고민해보시면 좋을 것 같아요 😄

hbkuk replied:
넵. 잘 읽었습니다 .. 👍

pch8388 reviewed (changes requested):
3단계 잘구현해주셨습니다 👏 
조금 더 수정해보았으면 합니다 🔥 
남겨드린 코멘트 참고하셔서 수정 후 리뷰요청 해주세요

hbkuk committed:
docs: ReadME update

hbkuk committed:
refactor: Section 엔티티의 역 혹은 구간 목록에 접근할 경우 정렬된 목록이 보장되도록 메서드 리팩토링

hbkuk committed:
fix: 정렬된 구간 목록을 반환하는 메서드에서 원본 객체는 수정하는 않도록 로직 수정

- refactor: 정렬된 구간 목록을 반환하는 메서드 명 수정

docs: ReadME update

hbkuk committed:
test: 정렬된 구간 목록을 반환하는 메서드에 대한 단위 테스트 추가

hbkuk committed:
refactor: 기존 정렬된 구간 목록을 반환하는 메서드를 Optional 라이브러리를 활용한 리팩토링

hbkuk committed:
refactor: 기존 Line Service에 몰려있는 의존성 분리(PathController, PathService) 및 관련 Mock Test 코드 분리

- fix: 불필요하게 'Station'이 붙은 파일명 수정

hbkuk committed:
refactor: 기존 Service 레이어에 정의된 Convert 메서드를 각 도메인별 Converter 클래스로 분리

hbkuk committed:
fix: PathFinderComponent에서 PathFinderService로 수정

- fix: 테스트 파일 이름을 PathServiceMockTest로 수정

- fix: 사용하지 않는 전역변수 삭제

hbkuk committed:
refactor: 기존 인수 테스트 클래스에서 사용된 데이터 타입을 전부 var 키워드로 대체하고 응답 값에서 ID를 추출하는 메서드를 숨김

- refactor: 기존 인수 테스트 클래스에 정의된 Convert 메서드를 LineSteps 클래스로 이동

- docs: LineSteps에 작성된 주석 삭제

hbkuk commented:
2단계, 3단계에서 리뷰해주신 내용을 최대한 반영해보았습니다. 😄 

혹시 놓친 부분이 있다면 코멘트 부탁드립니다. 😢

hbkuk commented on src/test/java/nextstep/subway/steps/LineSteps.java:
```diff
@@ -0,0 +1,113 @@
+import static nextstep.subway.utils.HttpResponseUtils.getCreatedLocationId;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class LineSteps {
```
`LineSteps` 뿐만 아니라 다른 `Steps`  대해서

전체적으로 리팩토링을 진행해보겠습니다. 😄

pch8388 commented on src/main/java/nextstep/subway/application/converter/LineConverter.java:
```diff
@@ -0,0 +1,24 @@
+
+import static nextstep.subway.application.converter.StationConverter.convertToStationResponses;
+
+public class LineConverter {
```
💯 
개인적으로도 개발할때 이런형태로 변환 객체를 두는 것을 선호하긴 하는데, 관리 포인트가 늘어날 수 있다는 점은 항상 생각해야 합니다

pch8388 commented on src/main/java/nextstep/subway/application/PathService.java:
```diff
@@ -0,0 +1,39 @@
+import static nextstep.subway.application.converter.PathConverter.convertToPathResponse;
+
+@Service
+public class PathService {
```
👍

pch8388 commented on src/test/java/nextstep/subway/unit/PathFinderServiceTest.java:
```diff
@@ -10,16 +9,14 @@
 
-@ApplicationTest
-public class PathFinderComponentTest {
+public class PathFinderServiceTest {
```
실제 클래스 이름도 PathFinder 라 같은 네이밍을 가지면 좋을것 같네요 😅

pch8388 commented on src/main/java/nextstep/subway/entity/Sections.java:
```diff
@@ -91,76 +118,62 @@ private Section createAndConnectNewSection(Section commonSection, Section nextSe
+    private void connectBasedOnSection(List<Section> sortedSections, int baseSectionIndex) {
+        Optional<Section> prependSection = findPrependSection(sortedSections);
+        if (isFindSection(prependSection)) {
+            moveSection(sortedSections, prependSection.get(), 0);
```
<img width="640" alt="스크린샷 2024-02-21 오후 8 46 17" src="https://github.com/next-step/atdd-subway-path/assets/17218212/d9f32737-70a0-451e-89e8-70e0a4f5712b">

Optional.get() 했을 때, 비어있는 Optional 이라면 `NoSuchElementException` 를 발생시킵니다.
비어있는 경우에 예외를 발생 시키고 싶다면 차라리 Optional 을 생성하지 말고 예외를 발생 시키는 것이 좋습니다 

`~~.findFirst().orElseThrow(~~~);`

Optional 은 nullable 하다는 것을 알리기 위한 것이기 때문에 null 일 경우에 뭘 할지에 대해서도 정의를 해주셔야 합니다 😅

hbkuk replied:
승철  리뷰어님.😄 

코멘트 감사드립니다!! 

<img width="439" alt="Optional" src="https://github.com/next-step/atdd-subway-path/assets/109803585/6cb6bd4e-235a-4cda-93cd-7f76bc9122e6">

다름이 아니라,
`findPrependSection` 메서드를 통해 **앞에 추가할 역**을 찾고, 아래와 같이 `isFindSection`을 통해 **실제로 역을 찾았는지 확인하는 것**이였는데요.

```
    private boolean isFindSection(Optional<Section> optionalSection) {
        return optionalSection.isPresent();
    }
```

혹시, `Optional` 라이브러리를 제가 잘못 알고 있는건지.. 궁금합니다. 😢

hbkuk replied:
또, Optional 객체를 이런 용도로 사용하는 것이 괜찮은지 궁금합니다.
(저도 많이 사용을 안해봐서 .. 다른 개발자 분들은 어떻게 느끼실지 궁금합니다 😄 )

pch8388 replied:
앗 아닙니다 제가 잘못봤네요 😢 
isPresent() 를 사용하신 부분을 확인못했네요 😅 
다만, Optional 은 한정적으로 사용해야 한다고 생각해요 🤔 

[참고](https://youtrack.jetbrains.com/issue/IDEA-194467) 의견도 한번 보시면 좋을 것 같아요

hbkuk replied:
넵. 감사합니다 😄

pch8388 commented on src/test/java/nextstep/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,151 @@
+     */
+    @DisplayName("지하철 노선을 생성한다.")
+    @Test
+    void 지하철_노선_생성() {
```
👍 
깔끔하게 구현되었네요
이런 식으로 시나리오 기반으로 작성하였을 때와 프로그래밍적인 요소가 들어간 테스트를 작성하였을 때의 장/단점에 대해 느끼셨으면 좋겠습니다

pch8388 commented on src/test/java/nextstep/subway/acceptance/PathAcceptanceTest.java:
```diff
@@ -42,11 +42,11 @@ public class PathAcceptanceTest {
      */
     @BeforeEach
-    public void setUp() {
+    public void 사전_노선_설정() {
```
💯 
이렇게 이름을 지어보겠다고 생각은 못해봤는데 깔끔하고 좋네요
다만, before 에서 setUp 이라는 이름을 쓰는 것은 관행적인 부분도 있기 때문에, 팀 구성원들 간 합의가 중요할 것 같습니다 😄

pch8388 reviewed (approved):
리팩토링 잘해주셨습니다 👍 
남겨드린 코멘트 참고해주세요 😄 
미션은 여기서 마무리하겠습니다.
추가로 궁금하신 점은 DM 부탁드립니다 🚀
